onnxruntime/core/providers/coreml/builders/impl/binary_op_builder.cc

```diff
@@ -36,7 +38,17 @@ Status BinaryOpBuilder::AddToModelBuilderImpl(ModelBuilder& model_builder, const
   std::unique_ptr<COREML_SPEC::NeuralNetworkLayer> layer = CreateNNLayer(model_builder, node);
 
   if (op_type == "Add") {
-    layer->mutable_add();
+    // original mutable_add() has limited broadcasting support
+    // updated to use CoreML- AddBroadcastableLayerParams which has more general broadcasting support
+    layer->mutable_addbroadcastable();
```

Comment:
Make everything by default broadcastable will make these ops running on CPU instead of ANE
Unofficial doc https://github.com/hollance/neural-engine/blob/master/docs/unsupported-layers.md#broadcastable-layers

Reply:
Thanks for the info! 

We should test to check though given that documentation is 3 years old. I would hope that CoreML was smart enough that if it can determine the layer doesn't actually need broadcasting it would use the non-broadcasting version and ANE, especially if coremltools was preferring creating models with the broadcastable layers. Seems like a super simple performance optimization to do.

Reply:
not sure if it's useful info, I tried to run perf_test with coreml ep on a simple Add graph (non broadcastable Add) 

and with `mutable_add()` vs. `mutable_addbroadcastable()`, the perf numbers are about the same though.

Is there any tools/approaches to directly check if these ops are running on CPU or ANE to further confirm?

Reply:
there's some more info about that from the same repo:
https://github.com/hollance/neural-engine/blob/master/docs/is-model-using-ane.md
though it hasn't been updated recently.

Reply:
I would expect we need a model with a lot of Add nodes to make sure there's a noticeable perf difference from ANE vs CPU.

Quick script to generate a model with Add/Sub/Div/Mul nodes (pick one) that do not require broadcasting. Specify data shape to make sure there is a significant amount of work to be done.

e.g. params to create a model with 100 Add nodes with shape {256, 256}  
`--op Add --num_nodes 100 --data_shape 256,256 Add100.onnx`

```py
import argparse
import numpy as np
import onnx
import os
import sys

from onnx import helper, numpy_helper, TensorProto

def make_model(op, num_nodes, data_shape):
    assert(num_nodes > 2)  # simplify code as first and last node need special handling

    cur_node = 0
    node_out = f'node_{cur_node}_out'
    nodes = [helper.make_node(op, inputs=['X', 'Y'], outputs=[node_out], name=f'node_{cur_node}')]
    cur_node += 1

    while cur_node < num_nodes:
        if cur_node == num_nodes - 1:
            new_node_out = 'Z'
        else:
            new_node_out = f'node_{cur_node}_out'

        nodes.append(helper.make_node(op, inputs=[node_out, 'W'], outputs=[new_node_out], name=f'node_{cur_node}'))
        node_out = new_node_out
        cur_node += 1

    # random numbers in initializer with no zeros so can safely be used with Div
    w_data = np.random.uniform(0.0001, 1.0, data_shape).astype('float32')

    model = helper.make_model(
        opset_imports=[helper.make_operatorsetid('', 15)],
        graph=helper.make_graph(
            name=f'{op}{num_nodes}',
            inputs=[helper.make_tensor_value_info('X', TensorProto.FLOAT, shape=data_shape),
                    helper.make_tensor_value_info('Y', TensorProto.FLOAT, shape=data_shape)],
            initializer=[numpy_helper.from_array(w_data, name='W')],
            outputs=[helper.make_tensor_value_info('Z', TensorProto.FLOAT, shape=data_shape)],
            nodes=nodes,
        ),
    )

    return model


def parse_args():
    parser = argparse.ArgumentParser(
        f"{os.path.basename(__file__)}:",
        description="""Create model with repeated nodes of Add/Div/Mul/Sub that are non-broadcasting."""
    )

    parser.add_argument("--op", type=str, default="Add", required=False, help="Add, Div, Mul or Sub")
    parser.add_argument("--num_nodes", type=int, default=50, required=False, help="Number of nodes")
    parser.add_argument("--data_shape",
                        type=lambda x: [int(i) for i in x.split(",")],
                        default="3,128,128",
                        required=False,
                        help="Shape to use for data. Provide comma separated list for the shape. "
                             "All values must be > 0. e.g. --data_shape 1,3,256,256",
    )
    parser.add_argument("output", type=str, help="Path to write model to.")
    return parser.parse_args()


if __name__ == '__main__':
    args = parse_args()
    model = make_model(args.op, args.num_nodes, args.data_shape)
    onnx.checker.check_model(model, True)
    onnx.save(model, args.output)
```

Reply:
can we detect the cases where we don't need broadcasting and use the non-broadcastable versions then?

Reply:
Maybe the Add on CPU is also efficient so ANE won't show too much gain here, you would need some context switch to show the performance gap, for example, one add followed by addbroadcastable, this may show the perf gap

But I guess using add instead of addbroadcastable when broadcast is not required is a good approach.